typescript/core/src/parse.ts

```diff
@@ -1,374 +1,415 @@
 import { crc32 } from "@foxglove/crc";
 
 import Reader from "./Reader";
-import { isKnownOpcode, MCAP_MAGIC, Opcode } from "./constants";
+import { MCAP_MAGIC, Opcode } from "./constants";
 import { McapMagic, TypedMcapRecord } from "./types";
 
 /**
  * Parse a MCAP magic string at `startOffset` in `view`.
  */
-export function parseMagic(
-  view: DataView,
-  startOffset: number,
-): { magic: McapMagic; usedBytes: number } | { magic?: undefined; usedBytes: 0 } {
-  if (startOffset + MCAP_MAGIC.length > view.byteLength) {
-    return { usedBytes: 0 };
+export function parseMagic(reader: Reader): McapMagic | undefined {
+  if (reader.bytesRemaining() < MCAP_MAGIC.length) {
+    return undefined;
   }
-  if (!MCAP_MAGIC.every((val, i) => val === view.getUint8(startOffset + i))) {
+  const magic = reader.u8ArrayBorrow(MCAP_MAGIC.length);
+  if (!MCAP_MAGIC.every((val, i) => val === magic[i])) {
     throw new Error(
       `Expected MCAP magic '${MCAP_MAGIC.map((val) => val.toString(16).padStart(2, "0")).join(
         " ",
-      )}', found '${Array.from(MCAP_MAGIC, (_, i) =>
-        view
-          .getUint8(startOffset + i)
-          .toString(16)
-          .padStart(2, "0"),
-      ).join(" ")}'`,
+      )}', found '${Array.from(magic, (_, i) => magic[i]!.toString(16).padStart(2, "0")).join(
+        " ",
+      )}'`,
     );
   }
-  return {
-    magic: { specVersion: "0" },
-    usedBytes: MCAP_MAGIC.length,
-  };
+  return { specVersion: "0" };
 }
 
 /**
- * Parse a MCAP record beginning at `startOffset` in `view`.
+ * Parse a MCAP record from the given reader
  */
-export function parseRecord({
-  view,
-  startOffset,
-  validateCrcs,
-}: {
-  view: DataView;
-  startOffset: number;
-  validateCrcs: boolean;
-}): { record: TypedMcapRecord; usedBytes: number } | { record?: undefined; usedBytes: 0 } {
-  if (startOffset + /*opcode*/ 1 + /*record content length*/ 8 >= view.byteLength) {
-    return { usedBytes: 0 };
+// eslint-disable-next-line @foxglove/no-boolean-parameters
+export function parseRecord(reader: Reader, validateCrcs = false): TypedMcapRecord | undefined {
+  const RECORD_HEADER_SIZE = 1 /*opcode*/ + 8; /*record content length*/
+  if (reader.bytesRemaining() < RECORD_HEADER_SIZE) {
+    return undefined;
   }
-  const headerReader = new Reader(view, startOffset);
+  const start = reader.offset;
+  const opcode = reader.uint8();
+  const recordLength = reader.uint64();
 
-  const opcode = headerReader.uint8();
-
-  const recordLength = headerReader.uint64();
   if (recordLength > Number.MAX_SAFE_INTEGER) {
     throw new Error(`Record content length ${recordLength} is too large`);
   }
+
   const recordLengthNum = Number(recordLength);
-  const recordEndOffset = headerReader.offset + recordLengthNum;
-  if (recordEndOffset > view.byteLength) {
-    return { usedBytes: 0 };
+
+  if (reader.bytesRemaining() < recordLengthNum) {
+    reader.offset = start; // Rewind to the start of the record
```

Comment:
this behavior of resetting the reader is interesting and should probably be called out in the doc comment on this function, and can we add unit tests for it?